File: calc.py
import csv

def calc_rm_anova(filename, cols):
    data = []
    with open(filename, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            data.append([float(row[c]) for c in cols])
    
    N = len(data)
    k = len(cols)
    
    grand_mean = sum(sum(row) for row in data) / (N * k)
    
    ss_total = sum(sum((val - grand_mean)**2 for val in row) for row in data)
    
    ss_subjects = k * sum(((sum(row)/k) - grand_mean)**2 for row in data)
    
    condition_means = [sum(data[i][j] for i in range(N))/N for j in range(k)]
    ss_conditions = N * sum((mean - grand_mean)**2 for mean in condition_means)
    
    ss_error = ss_total - ss_subjects - ss_conditions
    
    df_conditions = k - 1
    df_error = (N - 1) * (k - 1)
    
    ms_conditions = ss_conditions / df_conditions
    ms_error = ss_error / df_error
    
    F = ms_conditions / ms_error
    print(f"Cols {cols}: F = {F}")

calc_rm_anova('datasets/demo_all_analysis.csv', ['数学', '英語', '国語'])
calc_rm_anova('datasets/demo_all_analysis.csv', ['数学', '英語', '理科'])

File: calc.js
const fs = require('fs');
const path = require('path');

const file = fs.readFileSync('datasets/demo_all_analysis.csv', 'utf8');
const lines = file.trim().split('\n');
const headers = lines[0].split(',');

const colsToCalc = ['数学', '英語', '理科'];
const indices = colsToCalc.map(c => headers.indexOf(c));

const data = [];
for (let i = 1; i < lines.length; i++) {
    const parts = lines[i].split(',');
    data.push(indices.map(idx => parseFloat(parts[idx])));
}

const N = data.length;
const k = colsToCalc.length;

let grandSum = 0;
data.forEach(r => r.forEach(v => grandSum += v));
const grandMean = grandSum / (N * k);

let ssTotal = 0;
data.forEach(r => r.forEach(v => ssTotal += Math.pow(v - grandMean, 2)));

let ssSubjects = 0;
data.forEach(r => {
    const subMean = r.reduce((a, b) => a + b, 0) / k;
    ssSubjects += k * Math.pow(subMean - grandMean, 2);
});

let ssConditions = 0;
for (let j = 0; j < k; j++) {
    let condSum = 0;
    for (let i = 0; i < N; i++) {
        condSum += data[i][j];
    }
    const condMean = condSum / N;
    ssConditions += N * Math.pow(condMean - grandMean, 2);
}

const ssError = ssTotal - ssSubjects - ssConditions;
const dfConditions = k - 1;
const dfError = (N - 1) * (k - 1);

const msConditions = ssConditions / dfConditions;
const msError = ssError / dfError;

const F = msConditions / msError;

console.log(`msConditions=${msConditions}, msError=${msError}`);
console.log(`N=${N}, k=${k}`);
console.log(`F = ${F}`);
